### PR TITLE
GET UserPoint and User Trophy

### PR DESCRIPTION
**Pull request by Alzaif**
Had to change Docker version to 2.0 to run on my machine. May need a couple changes and "@CircuitBreaker".

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -1,4 +1,4 @@
-version: '3.7'
+version: '2.0'
```
Comment (fcgl) [inline docker-compose.yml]: What error do you get when it's 3.7?

File: src/main/java/com/fcgl/madrid/points/controller/UserPointsController.java [hunk 2]
```diff
@@ -14,13 +22,21 @@
...
     }
 
-    @RequestMapping(value = "/all", method = RequestMethod.GET)
+    @GetMapping("/all")
     public List<UserPoint> findAll() {
         return userPointsService.findAll();
     }
+
+    @GetMapping(value="/fallback")
+    public ResponseEntity<InternalStatus> failureWithFallback() {
```
Comment (fcgl) [inline src/main/java/com/fcgl/madrid/points/controller/UserPointsController.java]: This shouldn't go here, if you want to add this fallback endpoint you should create a `controller` in the `dev` package that you added

File: src/main/java/com/fcgl/madrid/points/controller/UserPointsController.java [hunk 3]
```diff
@@ -14,13 +22,21 @@
 public class UserPointsController {
```
Comment (fcgl) [inline src/main/java/com/fcgl/madrid/points/controller/UserPointsController.java:23]: Need to add a GET `userPoints` endpoint in this controller.
```
    @GetMapping("/userPoints")
    public List<UserPoint> getUserPoints(Long userId) {
        return userPointsService.getUserPoints(userId);
    }
```

File: src/main/java/com/fcgl/madrid/points/controller/UserTrophyController.java [hunk 4]
```diff
@@ -23,4 +32,13 @@ public void setPostService(UserTrophyService userTrophiesService) {
     public List<UserTrophy> findAll() {
         return userTrophiesService.findAll();
     }
+    // 
+    // public List<UserTrophy> findAll() {
+    //     return userTrophyRepository.findAll();
+    // }
+
+    @GetMapping(value="/fallback")
```
Comment (fcgl) [inline src/main/java/com/fcgl/madrid/points/controller/UserTrophyController.java]: Same as previous comment, this shouldn't go here. Should add a controller on `dev` package if you want to add this

File: src/main/java/com/fcgl/madrid/points/controller/UserTrophyController.java [hunk 5]
```diff
@@ -14,6 +21,8 @@
 public class UserTrophyController {
```
Comment (fcgl) [inline src/main/java/com/fcgl/madrid/points/controller/UserTrophyController.java:22]: Need to add a GET `userTrophies` endpoint in this controller.
```
    @GetMapping("/userTrophies")
    public List<UserPoint> getUserTrophies(Long userId) {
        return userTrophyService.getUserTrophies(Long userId);
    }
```

File: src/main/java/com/fcgl/madrid/points/service/UserPointsService.java [hunk 6]
```diff
@@ -19,4 +33,36 @@ public UserPointsService(UserPointsRepository userPointsRepository) {
     public List<UserPoint> findAll() {
         return userPointsRepository.findAll();
     }
+
+    // @CircuitBreaker(name = "backendA", fallbackMethod = "fallback")
```
Comment (fcgl) [inline src/main/java/com/fcgl/madrid/points/service/UserPointsService.java]: Missing function that calls `userPointsRepository.findByID`

File: src/main/java/com/fcgl/madrid/points/service/UserTrophyService.java [hunk 7]
```diff
@@ -19,4 +33,34 @@ public UserTrophyService(UserTrophyRepository userTrophyRepository) {
     public List<UserTrophy> findAll() {
         return userTrophyRepository.findAll();
     }
+
+    // @CircuitBreaker(name = "backendA", fallbackMethod = "fallback")
```
Comment (fcgl) [inline src/main/java/com/fcgl/madrid/points/service/UserTrophyService.java]: Missing function that calls `userTrophyRepository.findByID`

**Review by fcgl — CHANGES_REQUESTED** [review]
Good start, Missing some logic in the controller classes and the service classes.

File: docker-compose.yml (hunk 1 above)
Comment (Alzaif) [reply docker-compose.yml]: Been getting. Changing to 2.0 fixes it, I'll change it back to 3.7 for the sake of the PR.
`ERROR: Version in "./docker-compose.yml" is unsupported. You might be seeing this error because you're using the wrong Compose file version. Either specify a version of "2" (or "2.0")`

Comment (fcgl) [reply docker-compose.yml]: I think you might need to pull down a newer version of `docker-compose` and/or `docker`

File: src/main/java/com/fcgl/madrid/points/repository/UserPointsRepository.java [hunk 8]
```diff
@@ -2,9 +2,13 @@
...
 import com.fcgl.madrid.points.model.UserPoint;
 import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
 import org.springframework.stereotype.Repository;
+import java.util.List;
 
 @Repository
 public interface UserPointsRepository extends JpaRepository<UserPoint, Long> {
+  @Query("SELECT u FROM UserPoint u WHERE u.userId = ?1")
+    List<UserPoint> getUserPointsById(Long id);
```
Comment (fcgl) [inline src/main/java/com/fcgl/madrid/points/repository/UserPointsRepository.java]: There should only be one entry of UserPoint in the database. So this should be `UserPoint` instead of `List<UserPoint`

File: src/main/java/com/fcgl/madrid/points/model/UserPoint.java [hunk 9]
```diff
@@ -1,5 +1,5 @@
 package com.fcgl.madrid.points.model;
-
+
```
Comment (fcgl) [inline src/main/java/com/fcgl/madrid/points/model/UserPoint.java]: Not part of this PR. But noticed a bug when testing out the API.

On line 30 of this file `UserPoint.java` can you make the following change:
change
`this.totalPoints = tournamentPoints;`
to 
`this.tournamentPoints = tournamentPoints;`

File: src/main/java/com/fcgl/madrid/points/service/UserPointsService.java [hunk 10]
```diff
@@ -19,4 +33,37 @@ public UserPointsService(UserPointsRepository userPointsRepository) {
     public List<UserPoint> findAll() {
         return userPointsRepository.findAll();
     }
+
```
Comment (fcgl) [inline src/main/java/com/fcgl/madrid/points/service/UserPointsService.java]: Can we add some `Resilience4j` logic above this function please

Similar to this: https://github.com/fcgl/forum/pull/10/files#diff-40773a84dd74d64ebad9a4a76a122205R7, https://github.com/fcgl/forum/pull/10/files#diff-40773a84dd74d64ebad9a4a76a122205R122
```
@CircuitBreaker(name = "backendA", fallbackMethod = "fallback")
...
```

```
    /**
     *
     * @param ex Exception
     * @return ResponseEntity<InternalStatus>
     */
    private ResponseEntity<GetUserPointsResponse> fallback(Long userId, Exception ex) {
        String message = "Fallback: " + ex.getMessage();
        InternalStatus internalStatus = new InternalStatus(StatusCode.UNKNOWN, 500, message);
        GetUserPointsResponse postResponse = new GetUserPointsResponse(internalStatus, null);
        return new ResponseEntity<GetUserPointsResponse>(postResponse, HttpStatus.INTERNAL_SERVER_ERROR);
    }
```

Comment (fcgl) [reply src/main/java/com/fcgl/madrid/points/service/UserPointsService.java]: Can we also change the response type to be GetUserPointsResponse

That will be a class with these two attributes: UserPoint, InternalStatus

You can use this as reference

File: src/main/java/com/fcgl/madrid/points/service/UserTrophyService.java [hunk 11]
```diff
@@ -19,4 +33,38 @@ public UserTrophyService(UserTrophyRepository userTrophyRepository) {
...
+    }
+
+    // @CircuitBreaker(name = "backendA", fallbackMethod = "fallback")
+    // public ResponseEntity<InternalStatus>
+
+    /**
+    *@param
+    *@return
+    */
+    private ResponseEntity<InternalStatus> handleParamException(TransactionSystemException e) {
```
Comment (fcgl) [inline src/main/java/com/fcgl/madrid/points/service/UserTrophyService.java]: We want to stay away from having duplicate functions. Since this function as well as `getInternalStatusResponseEntity` is used twice I would recommend to add another class... Perhaps `ExceptionUtil` and add `handleParamException` and `getInternalStatusResponseEntity` there as public functions. Then import that class and call the necessary functions

**Review by fcgl — CHANGES_REQUESTED** [review]
* APIs work as expected
* Made a few small comments

File: .classpath [hunk 12]
```diff
@@ -0,0 +1,32 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (fcgl) [inline .classpath]: I think you pushed some stuff you shouldn't have? This file shouldn't be here I don't think

File: .project [hunk 13]
```diff
@@ -0,0 +1,23 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (fcgl) [inline .project]: I think you pushed some stuff you shouldn't have? This file shouldn't be here I don't think

File: .settings/org.eclipse.core.resources.prefs [hunk 14]
```diff
@@ -0,0 +1,5 @@
+eclipse.preferences.version=1
```
Comment (fcgl) [inline .settings/org.eclipse.core.resources.prefs]: I think you pushed some stuff you shouldn't have? This file shouldn't be here I don't think

File: .settings/org.eclipse.jdt.core.prefs [hunk 15]
```diff
@@ -0,0 +1,9 @@
+eclipse.preferences.version=1
```
Comment (fcgl) [inline .settings/org.eclipse.jdt.core.prefs]: I think you pushed some stuff you shouldn't have? This file shouldn't be here I don't think

File: bin/.gitignore [hunk 16]
```diff
@@ -0,0 +1,28 @@
+# Compiled class file
```
Comment (fcgl) [inline bin/.gitignore]: I think you pushed some stuff you shouldn't have? This file shouldn't be here I don't think

Anything inside of `bin/`

File: docker-compose.yml (hunk 1 above)
Comment (fcgl) [inline docker-compose.yml]: Change this back to `3.7`

File: src/main/java/com/fcgl/madrid/points/controller/UserPointsController.java [hunk 17]
```diff
@@ -14,13 +22,22 @@
...
     private UserPointsService userPointsService;
 
+    DevService devService;
+
+    @GetMapping("/userPoints")
+    public UserPoint getUserPointsById(Long userId) {
+    return userPointsService.getUserPointsById(userId);
```
Comment (fcgl) [inline src/main/java/com/fcgl/madrid/points/controller/UserPointsController.java]: Indentation if off here

File: src/main/java/com/fcgl/madrid/points/service/UserPointsService.java [hunk 18]
```diff
@@ -1,22 +1,59 @@
...
+	public UserPointsService(UserPointsRepository userPointsRepository) {
+		this.userPointsRepository = userPointsRepository;
+	}
+
+	@CircuitBreaker(name = "backendA", fallbackMethod = "fallback")
+	public List<UserPoint> findAll() {
+		return userPointsRepository.findAll();
+	}
+
+	public UserPoint getUserPointsById(Long userId) {
```
Comment (fcgl) [inline src/main/java/com/fcgl/madrid/points/service/UserPointsService.java]: The `@CircuitBreaker(name = "backendA", fallbackMethod = "fallback")` Should go above this function

Comment (fcgl) [inline src/main/java/com/fcgl/madrid/points/service/UserPointsService.java]: This should be returning `GetUserPointsResponse`

File: src/main/java/com/fcgl/madrid/points/service/UserTrophyService.java [hunk 19]
```diff
@@ -2,21 +2,41 @@
...
+
+	public List<UserTrophy> findAll() {
+		return userTrophyRepository.findAll();
+	}
 
-    @Autowired
-    public UserTrophyService(UserTrophyRepository userTrophyRepository) {
-        this.userTrophyRepository = userTrophyRepository;
-    }
+	public List<UserTrophy> getUserTrophiesById(Long userId) {
```
Comment (fcgl) [inline src/main/java/com/fcgl/madrid/points/service/UserTrophyService.java]: Missing `@CircuitBreaker(name = "backendA", fallbackMethod = "fallback")` above this function and the fallback method is needed in this class as well

Comment (fcgl) [reply src/main/java/com/fcgl/madrid/points/service/UserTrophyService.java]: You Should be returning something similar to `GetUserPointsResponse` here.  Maybe `GetUserTrophiesResponse` with `InternalStatus` and `List<UserTrophy>`

File: src/main/java/com/fcgl/madrid/points/controller/UserPointsController.java [hunk 20]
```diff
@@ -14,13 +22,22 @@
 public class UserPointsController {
     private UserPointsService userPointsService;
 
+    DevService devService;
+
+    @GetMapping("/userPoints")
+    public UserPoint getUserPointsById(Long userId) {
```
Comment (fcgl) [inline src/main/java/com/fcgl/madrid/points/controller/UserPointsController.java]: You're going to want to change this return type to be the POJO class that `getUserPointsById` returns once you update this PR. The same applies to the other endpoint you added for UserTrophies

**Review by fcgl — CHANGES_REQUESTED** [review]
* Few more comments
* There were some extra files that aren't needed

File: src/main/java/com/fcgl/madrid/points/service/UserPointsService.java (hunk 18 above)
Comment (Alzaif) [reply src/main/java/com/fcgl/madrid/points/service/UserPointsService.java]: Returning `GetUserPointResponse` for the `getUserPointById` method?

File: bin/.gitignore (hunk 16 above)
Comment (Alzaif) [reply bin/.gitignore]: My bad, removed those

File: src/main/java/com/fcgl/madrid/points/service/UserPointsService.java (hunk 18 above)
Comment (fcgl) [reply src/main/java/com/fcgl/madrid/points/service/UserPointsService.java]: Yeah, the `UserPoint` object is inside of `GetUserPointsResponse` right?

Comment (Alzaif) [reply src/main/java/com/fcgl/madrid/points/service/UserPointsService.java]: Yeah it is, just double checking before changing it

Comment (Alzaif) [reply src/main/java/com/fcgl/madrid/points/service/UserPointsService.java]: Something like this?
```

	@CircuitBreaker(name = "backendA", fallbackMethod = "fallback")
	public GetUserPointsResponse getUserPointsById(Long userId) {
		getUserPointResponse.setUserPoint(userPointsRepository.getUserPointsById(userId));
		return getUserPointResponse;
	}

```

File: bin/.gitignore (hunk 16 above)
Comment (fcgl) [reply bin/.gitignore]: This file needs to be removed as well, anything in bin

File: bin/.project [hunk 21]
```diff
@@ -0,0 +1,23 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (fcgl) [inline bin/.project]: This file needs to be removed

File: bin/.settings/org.eclipse.core.resources.prefs [hunk 22]
```diff
@@ -0,0 +1,5 @@
+eclipse.preferences.version=1
```
Comment (fcgl) [inline bin/.settings/org.eclipse.core.resources.prefs]: This file needs to be removed

File: bin/.settings/org.eclipse.jdt.core.prefs [hunk 23]
```diff
@@ -0,0 +1,9 @@
+eclipse.preferences.version=1
```
Comment (fcgl) [inline bin/.settings/org.eclipse.jdt.core.prefs]: This file needs to be removed

File: src/main/java/com/fcgl/madrid/points/model/response/GetUserTrophyResponse.java [hunk 24]
```diff
@@ -0,0 +1,44 @@
...
+import com.fcgl.madrid.points.model.UserTrophy;
+import com.fcgl.madrid.points.model.InternalStatus;
+
+import java.util.List;
+
+
+
+public class GetUserTrophyResponse {
+    private InternalStatus status;
+    private UserTrophy userTrophy;
```
Comment (fcgl) [inline src/main/java/com/fcgl/madrid/points/model/response/GetUserTrophyResponse.java]: `userTrophy` is not needed here, only List<UserTrophy>

File: src/main/java/com/fcgl/madrid/points/service/UserPointsService.java [hunk 25]
```diff
@@ -1,22 +1,61 @@
...
+
+import java.lang.Exception;
+import java.lang.Throwable;
+import java.lang.StringBuilder;
 
 @Service
 public class UserPointsService {
-    private UserPointsRepository userPointsRepository;
+	private UserPointsRepository userPointsRepository;
+	private GetUserPointsResponse getUserPointResponse;
```
Comment (fcgl) [inline src/main/java/com/fcgl/madrid/points/service/UserPointsService.java]: This shouldn't be here, it should be initialized inside of `getUserPointsById`

File: src/main/java/com/fcgl/madrid/points/service/UserPointsService.java [hunk 26]
```diff
@@ -1,22 +1,61 @@
...
+	}
+
+	public List<UserPoint> findAll() {
+		return userPointsRepository.findAll();
+	}
+
+	@CircuitBreaker(name = "backendA", fallbackMethod = "fallback")
+	public GetUserPointsResponse getUserPointsById(Long userId) {
+		getUserPointResponse.setUserPoint(userPointsRepository.getUserPointsById(userId));
+		return getUserPointResponse;
```
Comment (fcgl) [inline src/main/java/com/fcgl/madrid/points/service/UserPointsService.java]: As you currently have it `internalStatus` is set to null but this should not be the case.
This should be something along the lines of

```
UserPoint userPoint = getUserPointsById(userId);
return new GetUserPointsResponse(InternalStatus.OK, userPoint)
```

File: src/main/java/com/fcgl/madrid/points/service/UserTrophyService.java [hunk 27]
```diff
@@ -1,22 +1,62 @@
...
+
+import java.util.Set;
 import java.util.List;
+import java.util.ArrayList;
 
 @Service
 public class UserTrophyService {
-    private UserTrophyRepository userTrophyRepository;
+	private UserTrophyRepository userTrophyRepository;
+	private GetUserTrophyResponse getUserTrophyResponse;
```
Comment (fcgl) [inline src/main/java/com/fcgl/madrid/points/service/UserTrophyService.java]: This shouldn't be here, it should be initialized inside of `getUserTrophiesById `

File: src/main/java/com/fcgl/madrid/points/service/UserTrophyService.java [hunk 28]
```diff
@@ -1,22 +1,62 @@
...
+		this.userTrophyRepository = userTrophyRepository;
+	}
+
+	public List<UserTrophy> findAll() {
+		return userTrophyRepository.findAll();
+	}
+
+	public GetUserTrophyResponse getUserTrophiesById(Long userId) {
+		getUserTrophyResponse.setUserTrophies(userTrophyRepository.getUserTrophiesById(userId));
+		return getUserTrophyResponse;
```
Comment (fcgl) [inline src/main/java/com/fcgl/madrid/points/service/UserTrophyService.java]: As you currently have it `internalStatus` is set to null but this should not be the case.
This should be something along the lines of

```
List<UserTrophy> userTrophies = userTrophyRepository.getUserTrophiesById(userId);
return new GetUserTrophyResponse(InternalStatus.OK, userTrophies)
```

File: src/main/java/com/fcgl/madrid/points/service/UserTrophyService.java [hunk 29]
```diff
@@ -1,22 +1,62 @@
...
+
+	@Autowired
+	public UserTrophyService(UserTrophyRepository userTrophyRepository) {
+		this.userTrophyRepository = userTrophyRepository;
+	}
+
+	public List<UserTrophy> findAll() {
+		return userTrophyRepository.findAll();
+	}
+
```
Comment (fcgl) [inline src/main/java/com/fcgl/madrid/points/service/UserTrophyService.java:41]: `getUserTrophiesById` is Missing @CircuitBreaker annotation

**Review by fcgl — CHANGES_REQUESTED** [review]
Almost there, just a few more changes

File: src/main/java/com/fcgl/madrid/points/service/UserTrophyService.java [hunk 30]
```diff
@@ -1,22 +1,61 @@
...
-    @Autowired
-    public UserTrophyService(UserTrophyRepository userTrophyRepository) {
-        this.userTrophyRepository = userTrophyRepository;
-    }
+	/**
+	 *
+	 * @param ex Exception
+	 * @return ResponseEntity<InternalStatus>
+	 */
+	@CircuitBreaker(name = "backendA", fallbackMethod = "fallback")
```
Comment (fcgl) [inline src/main/java/com/fcgl/madrid/points/service/UserTrophyService.java]: You can delete this annotation. The only annotation should be in `getUserTrophiesById()`

File: src/main/java/com/fcgl/madrid/points/service/UserPointsService.java [hunk 31]
```diff
@@ -1,22 +1,59 @@
...
-    @Autowired
-    public UserPointsService(UserPointsRepository userPointsRepository) {
-        this.userPointsRepository = userPointsRepository;
-    }
+	/**
+	 *
+	 * @param ex Exception
+	 * @return ResponseEntity<InternalStatus>
+	 */
+	@CircuitBreaker(name = "backendA", fallbackMethod = "fallback")
```
Comment (fcgl) [inline src/main/java/com/fcgl/madrid/points/service/UserPointsService.java]: You can delete this annotation, you only need it in `getUserPointsById()`

**Review by fcgl — CHANGES_REQUESTED** [review]
This looks really good, just need to remove an annotation and this should be ready for testing on my side

File: src/main/java/com/fcgl/madrid/points/service/UserTrophyService.java [hunk 32]
```diff
@@ -1,22 +1,54 @@
...
+		this.userTrophyRepository = userTrophyRepository;
+	}
+
+	public List<UserTrophy> findAll() {
+		return userTrophyRepository.findAll();
+	}
+
+	@CircuitBreaker(name = "backendA", fallbackMethod = "fallback")
+	public GetUserTrophyResponse getUserTrophiesById(Long userId) {
+		return new GetUserTrophyResponse(InternalStatus.OK,	userTrophyRepository.getUserTrophiesById(userId));
```
Comment (fcgl) [inline src/main/java/com/fcgl/madrid/points/service/UserTrophyService.java]: This should be returning a `ResponseEntity< GetUserTrophyResponse >`.  I think that will get rid of the error

**Review by fcgl — CHANGES_REQUESTED** [review]
Getting the following error when making a request:
```
 java.lang.NoSuchMethodException: class com.fcgl.madrid.points.model.response.GetUserTrophyResponse class com.fcgl.madrid.points.service.UserTrophyService.fallback(class java.lang.Long,class java.lang.Throwable)
```

You can test if your endpoint works here: http://localhost:8083/swagger-ui.html#/

File: src/main/java/com/fcgl/madrid/points/service/UserPointsService.java [hunk 33]
```diff
@@ -1,22 +1,53 @@
...
+		this.userPointsRepository = userPointsRepository;
+	}
+
+	public List<UserPoint> findAll() {
+		return userPointsRepository.findAll();
+	}
+
+	@CircuitBreaker(name = "backendA", fallbackMethod = "fallback")
+	public GetUserPointsResponse getUserPointsById(Long userId) {
+		return  new GetUserPointsResponse(InternalStatus.OK, userPointsRepository.getUserPointsById(userId));
```
Comment (fcgl) [inline src/main/java/com/fcgl/madrid/points/service/UserPointsService.java]: This should be returning a ResponseEntity< GetUserPointsResponse >. I think that will get rid of this error: 

```
java.lang.NoSuchMethodException: class com.fcgl.madrid.points.model.response.GetUserPointsResponse class com.fcgl.madrid.points.service.UserPointsService.fallback(class java.lang.Long,class java.lang.Throwable)
```
